lib/web/service-worker/service-worker-container.js

```diff
@@ -0,0 +1,201 @@
+import { interceptor } from './interceptor.js'
+
+/**
+ * @typedef {Object} WorkerData
+ * @property {string} scriptUrl
+ * @property {unknown} options @todo
+ * @property {import('./client.js').SerializedClient} clientInfo
+ * @property {MessagePort} clientMessagePort
+ * @property {MessagePort} interceptorMessagePort
+ */
+
+const clientMessageChannel = new MessageChannel()
+const interceptorMessageChannel = new MessageChannel()
+
+/**
+ * @see https://w3c.github.io/ServiceWorker/#serviceworkercontainer-interface
+ */
+export class ServiceWorkerContainer {
+  /** @type {ServiceWorkerRegistration | undefined} */
+  #registration
+  /** @type {DeferredPromise<ServiceWorkerRegistration>} */
+  #ready
+
+  /** @todo Event handlers (oncontrollerchange, onmessage, onerror) */
+
+  constructor () {
+    this.#ready = new DeferredPromise()
+  }
+
+  /**
+   * @returns {Promise<ServiceWorkerRegistration>}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-ready
+   */
+  get ready () {
+    return this.#ready
+  }
+
+  /**
+   * @returns {ServiceWorker | null}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-controller
+   */
+  get controller () {
+    return this.#registration?.active || null
+  }
+
+  /**
+   * @param {string} scriptUrl
+   * @param {ServiceWorkerRegistrationOptions} [options]
+   * @returns {Promise<ServiceWorkerRegistration>}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-register
+   */
+  async register (scriptUrl, options = {}) {
+    /** @type {SerializedClient} */
+    const clientInfo = {
+      id: process.pid.toString(),
+      url: parseModuleUrlFromStackTrace(new Error()),
+      type: 'worker',
+      frameType: '???' /** @todo */
+    }
+
+    const worker = new Worker(new URL('./worker.ts', import.meta.url), {
+      name: `[worker ${scriptUrl}]`,
+      workerData: {
+        scriptUrl,
+        options,
+        clientInfo,
+        clientMessagePort: clientMessageChannel.port2,
+        interceptorMessagePort: interceptorMessageChannel.port2
+      },
+      transferList: [
+        clientMessageChannel.port2,
+        interceptorMessageChannel.port2
+      ]
+    })
+
+    const serviceWorker = this.#createServiceWorker(scriptUrl, worker)
+    const registration = new ServiceWorkerRegistration(serviceWorker)
+    this.#registration = registration
+
+    serviceWorker.addEventListener('statechange', () => {
+      if (serviceWorker.state === 'activating') {
+        this.#ready.resolve(registration)
+      }
+    })
+
+    this.#enableRequestInterception(interceptorMessageChannel)
+
+    return registration
+  }
+
+  /**
+   * @param {string} clientUrl
+   * @returns {Promise<ServiceWorkerRegistration | undefined>}
+   *
+   * @see https://w3c.github.io/ServiceWorker/#navigator-service-worker-getRegistration
+   */
+  getRegistration (clientUrl) {
```

Comment:
## Question

Should we abide by the rule that only _one_ Service Worker can control the same scope, like in the browser? Or should we allow an arbitrary number of Service Workers controlling the main thread? 

I lean toward the first option, would like to hear your thoughts on this as well.